airbyte-integrations/connectors/source-facebook-pages/source_facebook_pages/streams.py

```diff
@@ -0,0 +1,174 @@
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+#
+
+from abc import ABC
+from typing import Any, Iterable, Mapping, MutableMapping, Optional
+
+import pendulum
+import requests
+from airbyte_cdk.sources.streams.http import HttpStream
+from .metrics import PAGE_FIELDS, PAGE_METRICS, POST_FIELDS, POST_METRICS
+
+
+class FacebookStream(HttpStream, ABC):
+    url_base = "https://graph.facebook.com/v11.0/"
+    primary_key = "id"
+    data_field = "data"
+
+    def __init__(
+        self,
+        access_token: str = None,
+        page_id: str = None,
+        start_date: pendulum.datetime = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self._access_token = access_token
+        self._start_date = start_date
+        self._page_id = page_id
+
+    @property
+    def path_param(self):
+        return self.name[:-1]
+
+    def next_page_token(self, response: requests.Response) -> Optional[Mapping[str, Any]]:
+        return None
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        next_page_token = next_page_token or {}
+        params = {"access_token": self._access_token, **next_page_token}
+
+        return params
+
+
+class Page(FacebookStream):
+    """
+    API docs: https://developers.facebook.com/docs/graph-api/reference/page/,
+    """
+
+    def path(self, **kwargs) -> str:
+        if self._page_id:
+            return self._page_id
+
+    def request_params(
+        self,
+        stream_state: Mapping[str, Any],
+        stream_slice: Mapping[str, any] = None,
+        next_page_token: Mapping[str, Any] = None,
+    ) -> MutableMapping[str, Any]:
+        params = super().request_params(stream_state, stream_slice, next_page_token)
+        # we have to define which fields will return from Facebook API
+        # because FB API doesn't provide opportunity to get fields dynamically without delays
+        # so in PAGE_FIELDS we define fields that user can get from API
+        params["fields"] = PAGE_FIELDS
+
+        return params
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        yield response.json()
+
+
+class Post(FacebookStream):
+    """
+    API docs: https://developers.facebook.com/docs/graph-api/reference/post/,
+    """
+
+    def path(self, **kwargs) -> str:
+        if self._page_id:
+            return f"{self._page_id}/posts"
+
+    def parse_response(self, response: requests.Response, **kwargs) -> Iterable[Mapping]:
+        records = response.json().get(self.data_field) or []
```

Comment:
`records = response.json().get(self.data_field, [])`

Reply:
in case `self.data_field` is empty (None, False, etc) - it's better to have `[]` as a result